Migrate hooks and notification actions to typescript

larkox opened the pull request:
#### Summary
This PR is to clean the diff from https://github.com/mattermost/mattermost/pull/28334

We are migrating to typescript a couple of files and fixing some types along the way.

There is also a refactor on the desktop actions to make it easier later to force the notification

#### Ticket Link
NONE

#### Release Note
```release-note
NONE
```

streamer45 commented:
@larkox I thought the idea of using `code-reviewers` was to pick one of the two reviewers while retaining a manually assigned "subject expert". I am happy to review, of course, but I want to ensure the double selection was on purpose.

larkox commented:
@streamer45 You are right. Since the `code-reviewers` is adding two people, I defaulted a bit too much on it. I removed one of the reviewers assigned from `code-reviewers` and added a domain expert.

streamer45 reviewed (approved):
Great effort! To be on the safe side, I ran a quick QA on Calls Desktop notifications and everything looks good.

Side question: are `.tsx` tests a dream or is it something planned? :)

M-ZubairAhmed reviewed (approved):
I will thinking about this a couple of days ago, so this file isnt ts. And lo! and behold @larkox has a PR up

M-ZubairAhmed commented on webapp/channels/src/actions/websocket_actions.jsx:
The next beast is this file!

larkox commented:
@streamer45 Re: test files

Short answer: have to be done, but no real plan yet.

Long answer: This PR is just a support for a bigger PR, so I didn't want to get too tangled with all the details. That is why the hook types are not addressed in this PR needed. On the topic of test files, they are not so critical because they are not so prone to break on refactorings (one of the great things about having a type system). And at some point, we have to revisit all our tests and make them better. So I don't think it is not a priority migrating them now.

streamer45 commented:
> @streamer45 Re: test files
> 
> Short answer: have to be done, but no real plan yet.
> 
> Long answer: This PR is just a support for a bigger PR, so I didn't want to get too tangled with all the details. That is why the hook types are not addressed in this PR needed. On the topic of test files, they are not so critical because they are not so prone to break on refactorings (one of the great things about having a type system). And at some point, we have to revisit all our tests and make them better. So I don't think it is not a priority migrating them now.

Thanks @larkox . Mine was by no means a request but actually trying to understand myself if there's enough value in having typed tests in the first place or if they create unnecessary friction. 

For context, I am planning to do a general review across our org on test quality so you'll hear from me again on this as I see you have some good ideas!

larkox commented:
@streamer45 That sounds great. Actually, I tried to migrate the tests to TS yesterday, and I ended up in this rabbit hole: https://github.com/mattermost/mattermost/pull/28986

When that is merged, we can probably get those test in typescript more easily :stuck_out_tongue:

larkox commented:
@yasserfaraazkhan Not sure what plugins use these, but the idea is to test the following plugin hooks to make sure nothing has changed / broken:
- command will be posted
- message will be updated
- desktop notification hook

Also, you should check anything regarding notifications on the frontend side. Mainly that you still receive the notifications in the expected situations and it looks and sound ok.

yasserfaraazkhan commented:
/e2e-tests
